### PR TITLE
🧵 Enable threads by default

### PR DESCRIPTION
**Pull request by gsouquet**
The moment you've all been waiting for... Threads are enabled by default!

The labs flag had to change its key so that it will be enabled for all users by default and so that it overrides the previously set preferences.
We will keep it as a labs flag until the next release cycle. After that point it won't be possible to disable threads anymore and all references to it as a "beta feature" will be removed frm the product

Related to https://github.com/vector-im/element-meta/issues/566

Fixes https://github.com/vector-im/element-web/issues/19270
Fixes https://github.com/vector-im/element-web/issues/21910
Fixes https://github.com/vector-im/element-web/issues/23946

## Checklist

* [x] Tests written for new code (and old code if feasible)
* [x] Linter and other CI checks pass
* [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * 🧵 Enable threads by default ([\#9736](https://github.com/matrix-org/matrix-react-sdk/pull/9736)). Fixes vector-im/element-web#19270 vector-im/element-web#21910 and vector-im/element-web#23946.

### CODE REVIEW

**Comment by turt2live** [comment]
> https://github.com/orgs/matrix-org/teams/element-web-app-team could we bypass that check and merge that feature in. I'll DM you in time tomorrow to get this in. Thank you 🙏

By DM I think you mean message in the Web App Team room ;) 

(there's >1 person who can actively take ownership of these things now - it's important to use the team room)

**Comment by gsouquet** [comment]
@t3chguy I suggest removing the product review. 
This has already been discussed with @daniellekirkwood and she won't be able to give a review today.

@turt2live This is what I meant, hence why i tagged the app team above.

**Comment by t3chguy** [comment]
The issue it fixes is marked as needs product, and all changes like this need product sign off. There should be other product people aware of the project around, alternatively it shouldn't have been left to the last moment.

**Comment by amywalkerdev** [comment]
This approved from a product point of view, its been planned for a few weeks and both myself and @daniellekirkwood are aware of it

**Comment by t3chguy** [comment]
@amywalkerdev any chance you could leave an actual ✔️ - those have better visibility on github

**Review by amywalkerdev — APPROVED** [review]
Approved :)

File: src/MatrixClientPeg.ts [hunk 1]
```diff
@@ -233,7 +233,7 @@ class MatrixClientPegClass implements IMatrixClientPeg {
         opts.pendingEventOrdering = PendingEventOrdering.Detached;
         opts.lazyLoadMembers = true;
         opts.clientWellKnownPollPeriod = 2 * 60 * 60; // 2 hours
-        opts.experimentalThreadSupport = SettingsStore.getValue("feature_thread");
+        opts.experimentalThreadSupport = true;
```
Comment (t3chguy) [inline src/MatrixClientPeg.ts:236]: Shouldn't this still consider the labs flag (just renamed)?

File: src/components/structures/RoomSearchView.tsx [hunk 2]
```diff
@@ -109,12 +110,7 @@ export const RoomSearchView = forwardRef<ScrollPanel, Props>(
                                             );
                                         if (!bundledRelationship || event.getThread()) continue;
                                         const room = client.getRoom(event.getRoomId());
-                                        const thread = room.findThreadForEvent(event);
-                                        if (thread) {
-                                            event.setThread(thread);
-                                        } else {
-                                            room.createThread(event.getId(), event, [], true);
-                                        }
+                                        room.createThread(event.getId(), event, [], true);
```
Comment (t3chguy) [inline src/components/structures/RoomSearchView.tsx:113]: This won't do `event.setThread` if the thread already exists - are we certain something else will?

File: src/components/views/context_menus/MessageContextMenu.tsx [hunk 3]
```diff
@@ -71,11 +71,11 @@ const ReplyInThreadButton = ({ mxEvent, closeMenu }: IReplyInThreadButton) => {
     if (Boolean(relationType) && relationType !== RelationType.Thread) return null;
 
     const onClick = (): void => {
-        if (!localStorage.getItem("mx_seen_feature_thread")) {
-            localStorage.setItem("mx_seen_feature_thread", "true");
+        if (!localStorage.getItem("mx_seen_feature_threadstable")) {
```
Comment (t3chguy) [inline src/components/views/context_menus/MessageContextMenu.tsx:74]: It looks like you killed off the MessageActionBar bits of this, so doubt it'll work with only half the bits kept.

File: src/components/views/rooms/EventTile.tsx [hunk 4]
```diff
@@ -496,22 +496,11 @@ export class UnwrappedEventTile extends React.Component<EventTileProps, IState>
...
-         * that will be solved when there are proper backend support for threads
-         * We currently have no reliable way to discover than an event is a thread
-         * when we are at the sync stage
-         */
-        if (!thread) {
-            const room = MatrixClientPeg.get().getRoom(this.props.mxEvent.getRoomId());
-            thread = room?.findThreadForEvent(this.props.mxEvent);
-        }
-        return thread ?? null;
+        return this.props.mxEvent.getThread() ?? null;
```
Comment (t3chguy) [inline src/components/views/rooms/EventTile.tsx:503]: Will this not re-introduce the race condition if the user's HS doesn't support the backend support yet, e.g. Dendrite?

File: src/settings/Settings.tsx [hunk 5]
```diff
@@ -255,57 +255,13 @@ export const SETTINGS: { [setting: string]: ISetting } = {
...
-                    </>
-                ),
-            feedbackLabel: "thread-feedback",
-            feedbackSubheading: _td(
-                "Thank you for trying the beta, " + "please go into as much detail as you can so we can improve it.",
-            ),
-            image: require("../../res/img/betas/threads.png"),
-            requiresRefresh: true,
-        },
+        default: true,
```
Comment (t3chguy) [inline src/settings/Settings.tsx:264]: You seem to have removed the Beta info, won't this prevent users from being able to disable this without having to tweak config.json?>

File: test/components/structures/TimelinePanel-test.tsx [hunk 6]
```diff
@@ -163,12 +163,13 @@ describe("TimelinePanel", () => {
             expect(readMarkersSent).toEqual(["ev1"]);
         });
 
-        it("sends public read receipt when enabled", () => {
+        it("sends public read receipt when enabled", async () => {
```
Comment (t3chguy) [inline test/components/structures/TimelinePanel-test.tsx:166]: Does this need be async?

File: test/components/views/settings/tabs/user/__snapshots__/LabsUserSettingsTab-test.tsx.snap [hunk 7]
```diff
@@ -67,78 +67,6 @@ exports[`<SecurityUserSettingsTab /> renders settings marked as beta as beta car
       </div>
     </div>
   </div>
-  <div
```
Comment (t3chguy) [inline test/components/views/settings/tabs/user/__snapshots__/LabsUserSettingsTab-test.tsx.snap:70]: This shows that users won't be able to disable the default-on, making it closer to forced-on which wasn't the plan for this release

**Review by t3chguy — CHANGES_REQUESTED** [review]

File: src/components/views/rooms/EventTile.tsx (hunk 4 above)
Comment (gsouquet) [reply src/components/views/rooms/EventTile.tsx:503]: I did not consider dendrite in the mix, good shout 👍

Comment (t3chguy) [reply src/components/views/rooms/EventTile.tsx]: No need it be even Dendrite, we often get rageshakes from servers without the /hierarchy API, those are over a year old

**Review by t3chguy — APPROVED** [review]

**Pull request merged**